Gpii 2966 Initial review of endpoints being added

sgithens opened the pull request:
Hi @cindyli . There is some cleanup, plus docs and tests needed for this still, but I wanted to at least run the endpoints by you before the weekend and get any comments on them in general.  Also there are 3 new couch views for this.

cindyli commented on testData/dbData/views.json:
```diff
@@ -10,6 +10,15 @@
             "findAuthorizationByAccessToken": {
                 "map": "function(doc) {if (doc.type === 'gpiiAppInstallationAuthorization' && doc.revoked === false) {emit(doc.accessToken, {'_id': doc.clientId, 'authorization': doc})}}"
+            },
+            "listKeysForPrefsSafe": {
```
Most view names starts with "find" and these 2 use "list". Is there a reason in using a different word? If not, can you sync up the view names?

sgithens replied:
I think this makes sense @cindyli . I will probably leave `listPrefsSafes` as it is, since there is not search argument given to it, it really is just a listing.

cindyli commented on testData/dbData/views.json:
```diff
@@ -10,6 +10,15 @@
+                "map": "function(doc) {if (doc.type === 'gpiiKey') emit(doc.prefsSafeId, null) }"
+            },
+            "listPrefsSafes": {
+                "map": "function(doc) { if (doc.type === 'prefsSafe') emit(doc._id, {  'name': doc.name, 'email': doc.email,   'created': doc.timestampCreated,  'updated': doc.timestampUpdated })}"
```
Remove extra spaces before "created" and "updated".

cindyli commented on gpii/node_modules/preferencesServer/src/prefsSafesHandlers.js:
```diff
@@ -0,0 +1,129 @@
+});
+
+gpii.preferencesServer.listKeysForPrefsSafe.handler.getKeysForPrefsSafe = function (prefsService, preferencesServer, request, prefsSafeId, view) {
+    var finalDirectModel = fluid.extend(true, {}, prefsService.dataStore.listKeysForPrefsSafeDataSource.options.directModel, {
```
I don't see a value for `prefsService.dataStore.listKeysForPrefsSafeDataSource.options.directModel`? Would it be more straightforward to change this line to:

```
var directModel = {
    prefsSafeId: prefsSafeId
};
```

cindyli commented on gpii/node_modules/preferencesServer/src/prefsSafesHandlers.js:
```diff
@@ -0,0 +1,129 @@
+        prefsSafeId: prefsSafeId});
+    var prefsPromise = prefsService.dataStore.listKeysForPrefsSafeDataSource.get(finalDirectModel);
+    prefsPromise.then(function (data) {
+        var togo = [];
```
line 76-81 can be simplified using `fluid.transform`.

cindyli commented on gpii/node_modules/preferencesServer/src/prefsSafesHandlers.js:
```diff
@@ -0,0 +1,129 @@
+
+gpii.preferencesServer.prefsSafeWithKeysGet.handler.get = function (prefsService, preferencesServer, request, prefsSafeId, view) {
+    // TODO Refactor from above, copied/pasted
+    var finalDirectModel = fluid.extend(true, {}, prefsService.dataStore.listKeysForPrefsSafeDataSource.options.directModel, {
```
Same comment as above on the use of `prefsService.dataStore.listKeysForPrefsSafeDataSource.options.directModel`. This type of comment applies across the pull request.

cindyli commented on gpii/node_modules/preferencesServer/src/prefsSafesHandlers.js:
```diff
@@ -0,0 +1,129 @@
+    }
+});
+
+gpii.preferencesServer.prefsSafeWithKeysGet.handler.get = function (prefsService, preferencesServer, request, prefsSafeId, view) {
```
This function does 2 things:
1. Find keys associated with a prefs safe;
2. Find the data of this prefs safe and add keys retrieved at step 1 to the data.

A couple of suggestions:
1. The step 1 is shared by this handler and `gpii.preferencesServer.listKeysForPrefsSafe.handler`, so it can be extracted into a function to get rid of the duplication;
2. Use [`fluid.promise.fireTransformEvent()`](https://docs.fluidproject.org/infusion/development/PromisesAPI.html#fluidpromisefiretransformeventevent-payload-options) to chain these 2 promises that the output of the first promise can be passed to the second promise. Using `fluid.promise.fireTransformEvent()` also helps to reuse the function in the first suggestion.

cindyli commented on gpii/node_modules/preferencesServer/src/prefsSafesHandlers.js:
```diff
@@ -0,0 +1,129 @@
+    // TODO Refactor from above, copied/pasted
+    var finalDirectModel = fluid.extend(true, {}, prefsService.dataStore.listKeysForPrefsSafeDataSource.options.directModel, {
+        prefsSafeId: prefsSafeId});
+    var prefsPromise = prefsService.dataStore.listKeysForPrefsSafeDataSource.get(finalDirectModel);
```
All handlers in this script are accessing `prefsService.dataStore` directly, which is not recommended. The `prefsService` acts as a bridge that helps to decouple the database operation from other GPII modules including preferencesServer. `prefsService` should provide functions that make sense to request handlers instead of having request handlers directly deal with data that are in raw document structure. I would suggest to move the 2 steps handling in this function to `prefsService`.

This comment applies to other handlers in this script.

cindyli commented:
Nice work, @sgithens. I added "needs work" label as you will continue to add tests and docs. Once those are out of way, please remove this label. Thanks.

gtirloni commented:
ok to test

sgithens commented:
@cindyli The only test failure from the build is 1 test from pouchdb manager which I know is being worked on.  In the meantime if you wanted to give this another review when you have time that would be great.

cindyli commented:
@sgithens, [@the-t-in-rtf's recent update on gpii-pouchdb](https://github.com/GPII/universal/pull/705#issuecomment-435426104) should fix this CI failure.

cindyli commented on documentation/DataModel.md:
```diff
@@ -6,3 +6,123 @@ configuration, CouchDB-compatible PouchDB is used for the data storage.
+### Key-in Documents
+
+When users key-in to the GPII using a USB stick, NFC card, or other mechanism, the unique `gpiiKey` on the device will
+be matched to the CouchDB `id` on a document with type `gpiiKey`. This document contains the important fields `prefsSafeId`
```
CouchDB `id` -> CouchDB `_id`. Missing underscore.

sgithens replied:
Fixed.

cindyli commented on documentation/DataModel.md:
```diff
@@ -6,3 +6,123 @@ configuration, CouchDB-compatible PouchDB is used for the data storage.
+        "type": "gpiiCloudSafeCredential",
+        "schemaVersion": "0.1",
+        "prefsSafeId": "prefsSafe-7",
+        "gpiiExpressUserId": "org.couch.db.user:prefs7user",
```
The corresponding "name" value at line 101 is "prefs7user". Why does the value saved here have a prefix "org.couch.db.user:"?

sgithens replied:
This is because the field is managed by gpii-express-user which uses that prefix. I've added a comment in the markdown noting this.

cindyli commented on documentation/DataModel.md:
```diff
@@ -6,3 +6,123 @@ configuration, CouchDB-compatible PouchDB is used for the data storage.
+        "type": "user",
+        "email": null,
+        "roles": [],
+        "username": "prefs7user",
```
If names/passwords are saved in a document type from `gpii-express-user`, shall "name" and "password" fields defined in the "user" document type (line 32, 33 above) be removed?

sgithens replied:
Good point. The password is definately not needed, but we may want to keep Name as a piece of metadata that stays around in case username/password logins are added or removed.  For now I've removed password from the docs here.

cindyli commented on gpii/node_modules/gpii-db-operation/src/DbDataStore.js:
```diff
@@ -97,6 +97,48 @@ fluid.defaults("gpii.dbOperation.dbDataStore", {
+            type: "gpii.dbOperation.dbDataSource",
+            options: {
+                requestUrl: "/_design/views/_view/listPrefsSafes",
+                termMap: {
```
remove line 118-120

sgithens replied:
Fixed.

cindyli commented on gpii/node_modules/gpii-db-operation/src/DbDataStore.js:
```diff
@@ -97,6 +97,48 @@ fluid.defaults("gpii.dbOperation.dbDataStore", {
+            options: {
+                requestUrl: "/_design/views/_view/expressUserSafeLoginLookup?include_docs=true&key=%22%userRecordId%22",
+                termMap: {
+                    userRecordId: "%userRecordId"
```
It would be nice to renamed the field name here ("userRecordId") with the actual field name "gpiiExpressUserId" defined in the document type.

sgithens replied:
Fixed.

cindyli commented on gpii/node_modules/gpii-db-operation/src/DbDataStoreUtils.js:
```diff
@@ -74,6 +74,37 @@ gpii.dbOperation.dbDataStore.findRecord = function (dataSource, directModel, val
+ * record is not empty.
+ * @return {Promise} A promise for the retrieved record.
+ */
+gpii.dbOperation.dbDataStore.findRecordList = function (dataSource, directModel, dataProcessFunc) {
```
I don't see a case where `dataProcessFunc` is provided and used. This argument should be removed if it's not in use.

sgithens replied:
For the `dataProcessFunc`, I added it to match the similar method `findRecord` because it seems useful for flexible processing.  From grepping the code base though, it doesn't look like anyone is exercising it when using `findRecord` either. 

The reason that I added the `findRecordList` in addition to `findRecord`, is that it is meant to always return a list, whereas `findRecord` only returns a single item. It's also slightly different when processing as well, because `findRecordList` can be empty as an acceptable value, whereas `findRecord` has special processing whenever something is empty to make it undefined.

cindyli commented on gpii/node_modules/preferencesServer/src/cloudSafeCredHandlers.js:
```diff
@@ -0,0 +1,66 @@
+            request.events.onSuccess.fire(data);
+        },
+        function (/* err */) {
+            request.res.status(401);
```
The typical way for line 59-63 is: 
```
        request.events.onError.fire({
            isError: true,
            message: "Unable to unlock a Preferences Safe with the supplied credentials.",
            statusCode: 401
        });
```
The `statusCode` fired with the onError will automatically sets the responded http status code to 401.

sgithens replied:
Thanks, have moved it to our standard idiom!

cindyli commented on gpii/node_modules/preferencesServer/test/cloudSafeCredTests.js:
```diff
@@ -0,0 +1,300 @@
+
+gpii.tests.preferencesServer.cloudSafeCred.put.successFixtures = [
+    {
+        name: "PUT: Add a cloudsafe credential to an existing prefssafe",
```
Would be nice to test failed cases such as when the given prefs safe id doesn't exist or the given user name or password is invalid.

sgithens replied:
Added a test (and fixed the implementation!) for when the prefs safe does not exist. This is a test in the unlock section for the invalid username and password case. It uses the same semantics as successful tests with a json payload returned, so it's in there.

cindyli commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -73,6 +108,41 @@ var fluid = fluid || require("infusion");
                 namespace: "updateGpiiKey"
+            }],
+            onCloudSafeCredCreate: [{
+                func: "{dataStore}.findById",
```
When prefs safe id is not found, the operation should stop with an error.

sgithens replied:
This has been fixed.

cindyli commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -73,6 +108,41 @@ var fluid = fluid || require("infusion");
+                namespace: "findById"
+            },
+            {
+                func: "{expressUserUtils}.createNewUser",
```
Curious what would happen if the given user name already exists. Can you add a test case for it too? Thanks.

the-t-in-rtf replied:
The underlying API will return an error, but it's good to test the handling and relaying of that error in this context.

sgithens replied:
I've added a test for this, and also added a message from the prefs server adding more detail to the couch conflict error raised by gpii-express-user.

cindyli commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -83,7 +153,60 @@ var fluid = fluid || require("infusion");
+     *
+     * @param {Object} that - Instance of `gpii.preferencesServer.preferencesService`.
+     * @param {Object} options - Data used to create the new credentials.
+     * @param {Object} options.prefsSafeId - Id for the preferences safe we are adding
```
Parameters stated at line 167-169 are strings instead of objects.

sgithens replied:
Fixed.

cindyli commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -83,7 +153,60 @@ var fluid = fluid || require("infusion");
+     *
+     * @param {Object} that - Instance of `gpii.preferencesServer.preferencesService`.
+     * @param {Object} options - Data used to unlock the safe.
+     * @param {Object} options.username - Username to unlock the Preferences Safe.
```
Parameters described at line 189, 190 are strings too.

sgithens replied:
Fixed.

cindyli commented on gpii/node_modules/preferencesServer/test/prefsSafesTests.js:
```diff
@@ -0,0 +1,373 @@
+    };
+};
+
+gpii.tests.preferencesServer.prefsSafes.get.successFixtures = [
```
Please test failed cases too. This comment applies to all new tests.

cindyli commented on gpii/node_modules/preferencesServer/test/preferencesServiceTests.js:
```diff
@@ -0,0 +1,437 @@
+fluid.require("%gpii-universal");
+gpii.loadTestingSupport();
+
+fluid.registerNamespace("gpii.tests.preferencesServer.preferencesService");
```
This file converts the preferencesService tests from in-browser tests to node tests, but I don't see new tests for new APIs like cloudSafeCredCreate() and cloudSafeUnlock().

sgithens replied:
Thanks for catching this.  I added tests for the new endpoints that use these methods, but not for direct usage of the methods.

cindyli commented on tests/all-tests.js:
```diff
@@ -69,6 +69,8 @@ var testIncludes = [
     "../gpii/node_modules/ontologyHandler/test/node/OntologyHandlerTests.js",
     "../gpii/node_modules/pouchManager/test/pouchManagerTests.js",
     "../gpii/node_modules/preferencesServer/test/preferencesServerTests.js",
+    "../gpii/node_modules/preferencesServer/test/cloudSafeCredTests.js",
```
Add preferencesServiceTests.js too.

sgithens replied:
Oh yes, it would help if I added the test cases I had just converted!

cindyli commented:
Update [.nycrc](https://github.com/sgithens/universal/blob/GPII-2966/.nycrc) to reflect changes on files.

cindyli commented on gpii/node_modules/preferencesServer/test/prefsSafesTests.js:
```diff
@@ -0,0 +1,373 @@
+
+
+
+/////////////////////// TESTING prefsSafes PUT (Save) ////////////////////////////
```
Missing tests for:
* /prefssafe/:prefsSafeId PUT
* /prefssafe-key-create POST

sgithens commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -83,7 +153,60 @@ var fluid = fluid || require("infusion");
+     * 2. If successful, take the gpii-express-user.id and look up a cloud credentials doc for it.
+     * 3. Fetch the Safe for that and return it.
+     *
+     * In the future the number of couch calls may by optimized.
```
Added TODO.

sgithens commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -83,7 +153,60 @@ var fluid = fluid || require("infusion");
+     * Handler for creating cloud safe credentials.
+     *
+     * This handler will first fetch the preferences safe to make sure it
+     * exists and it active.  It will then take the username and password and
```
fixed.

cindyli commented on documentation/DataModel.md:
```diff
@@ -6,3 +6,123 @@ configuration, CouchDB-compatible PouchDB is used for the data storage.
+        "revoked": false,
+        "revokedReason": null,
+        "timestampCreated": "2017-12-14T19:55:11.641Z",
+        "timestampUpdated": null
```
Add the field `timestampRevoked`.

sgithens replied:
This is updated.

cindyli commented on documentation/DataModel.md:
```diff
@@ -6,3 +6,123 @@ configuration, CouchDB-compatible PouchDB is used for the data storage.
+        "schemaVersion": "0.1",
+        "prefsSafeId": "prefsSafe-7",
+        "gpiiExpressUserId": "org.couch.db.user:prefs7user",
+        "_id": "8f3085a7-b65b-4648-9a78-8ac7de766997"
```
Move the line for "_id" to the beginning of the JSON data block to match the convention above. The same comment applies to the JSON block below with "_id" at the bottom at line 112.

sgithens replied:
This is updated.

cindyli commented on gpii/node_modules/gpii-db-operation/src/DbUtils.js:
```diff
@@ -36,3 +36,38 @@ gpii.dbOperation.composeError = function (error, termMap) {
+ * The current criteria for validity is that the item is not undefined and has
+ * the `type` that is supplied in the arguments.
+ *
+ * @param  {Object} item         Item, mostly a return from another API or CouchDB call
```
Follow the JSDoc convention. Replace spaces in between "item" and "Item" with "-". The same comment applies to other lines.

sgithens replied:
This should be fixed now.

cindyli commented on gpii/node_modules/preferencesServer/src/preferencesServer.js:
```diff
@@ -52,6 +52,52 @@ fluid.defaults("gpii.preferencesServer", {
             type: "gpii.preferencesServer.put.handler"
+        },
+        // GPII-2966
+        prefsSafeGet: {
```
Update documentation with these new endpoints.

sgithens replied:
All the endpoints should be in the doco now.

cindyli commented on gpii/node_modules/gpii-db-operation/src/DbDataStore.js:
```diff
@@ -167,6 +221,22 @@ fluid.defaults("gpii.dbOperation.dbDataStore", {
             ]
             // id
         },
+        findPrefsSafeList: {
```
Add tests for these new API functions.

sgithens replied:
Added to dbdatastore tests.

cindyli commented on testData/dbData/views.json:
```diff
@@ -11,6 +11,18 @@
+            "listPrefsSafes": {
+                "map": "function(doc) {if (doc.type === 'prefsSafe') emit(doc._id, {'name': doc.name, 'email': doc.email, 'created': doc.timestampCreated, 'updated': doc.timestampUpdated})}"
+            },
+            "findPrefsSafeByName": {
```
This view doesn't seem to be used.

sgithens replied:
Thanks for catching this, it was from earlier in the gpii express integration work, and did not end up being used.  Removed.

cindyli commented on testData/dbData/views.json:
```diff
@@ -11,6 +11,18 @@
                 "map": "function(doc) {if (doc.type === 'gpiiAppInstallationAuthorization' && doc.revoked === false) {emit(doc.accessToken, {'_id': doc.clientId, 'authorization': doc})}}"
             },
+            "findKeysForPrefsSafe": {
+                "map": "function(doc) {if (doc.type === 'gpiiKey' || doc.type === 'gpiiCloudSafeCredential') emit(doc.prefsSafeId, null)}"
```
This view returns not only GPII keys but also safe credentials. Is this intentional?

sgithens replied:
It is, and ideally I want it to return any future related docs and credential types we come up with, such as Nova tokens or anything else.  Renamed to findRelatedDocsForPrefsSafe

sgithens commented:
@cindyli Hi Cindy!  I believe this should be ready for another round of review.  Thanks!

cindyli commented on documentation/PreferencesServer.md:
```diff
@@ -349,6 +349,202 @@ There are two important things to note here:
    to `http://registry.gpii.net/common/fontSize`. Since we do not allow the same setting to be present multiple times in
    the NP set, the fontSize has been stored in the flat ontology and removed from the ISO24751 block.
 
+### GET /prefssafe/:prefsSafeId
```
For each endpoint, it would be helpful to add examples of a query URL, input and output (success or fail) payloads.

sgithens replied:
Hi! I believe these should all be added now.

cindyli commented on documentation/PreferencesServer.md:
````diff
@@ -349,6 +349,202 @@ There are two important things to note here:
+```json
+{
+    "prefsSafe": {
+        "id": "prefsSafeId",
````
I think it might be more explanatory to use real data in examples. Also expand omitted blocks like line 371, 373. Let me know what you think. Thanks.

cindyli commented on documentation/PreferencesServer.md:
````diff
@@ -349,6 +349,202 @@ There are two important things to note here:
+}
+```
+
+If the username and password do not match a record the following error is returned:
````
If there could be different error responses, explain them all. Thanks.

sgithens replied:
I put in some notes about error payloads for superfluous errors.

sgithens commented on documentation/DataModel.md:
```diff
@@ -6,3 +6,124 @@ configuration, CouchDB-compatible PouchDB is used for the data storage.
 
 The details of the GPII data model can be found [here](https://wiki.gpii.net/w/Keys,_KeyTokens,_and_Preferences).
+
+## Preference Safes Overview
```
This sounds great. I've created this ticket to track the work, and the discussion we can hopefully get on the agenda for it next during the arch meeting.

https://issues.gpii.net/browse/GPII-3694

sgithens commented:
@cindyli  I made the documentation updates from above... I think this should be ready for another round of review.  Thanks!

cindyli commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -223,6 +486,47 @@ var fluid = fluid || require("infusion");
+            });
+        }
+        else {
+            dataStore.addGpiiKey({
```
Adding a key without verifying the existence of prefsSafeId and prefsSetId will result in the unsynced data in "gpiiKey" and "prefsSafe" documents.

sgithens replied:
Good call, I added a check for this, and unit tests for it.

cindyli commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -73,6 +125,45 @@ var fluid = fluid || require("infusion");
+                func: "gpii.dbOperation.verifyExists",
+                args: ["{arguments}.0", "prefsSafe", "The prefs safe does not exist"]
+            }, {
+                func: "{expressUserUtils}.createNewUser",
```
What if the prefs safe already has an existing username? Shall this func [check the existing user account](https://github.com/GPII/gpii-express-user/blob/master/src/js/server/signup.js#L21) before creating a new one?

sgithens replied:
I think we should hold off for now on adding any type of limitation like that.  I think part of the beauty of our credential/key document structure is that you can have multiple keys/credentials of various types, and choose to revoke them, delete them, have multiple legit ones at the same time.  As of now, this setup is compatible with how gpii-express-user works, and we could have more than 1 legitimate username/password login for a prefssafe if a use case for that presented itself.

I imagine as the "Morphic World" UI design moves along, and someone works through the raminifications of deployments we might have more guidance on this, but I'd like to leave it flexible for now.

cindyli replied:
Thanks for the explanation. It makes sense.

cindyli commented on documentation/DataModel.md:
```diff
@@ -6,3 +6,124 @@ configuration, CouchDB-compatible PouchDB is used for the data storage.
+        "schemaVersion": "0.1",
+        "prefsSafeType": "user",
+        "name": null,
+        "email": null,
```
It just came to me that, with the data model change (the removal of "password" field from "prefsSafe" document type and the addition of "timestampRevoked" into "gpiiKey" type), other relevant places in the repo should be changed accordingly. This includes [the conversion script](https://github.com/GPII/universal/blob/master/scripts/shared/prefsSetsDbUtils.js#L73) that converts preferences json files at testData/preferences to couchDB document structures as well as other data used for unit and integration tests. Doing a search using keywords like "password", "gpiiKey" will reveal more.

sgithens replied:
Good point.  I've removed all the old `password` fields from everywhere: code, scripts, and tests.  I'm not familiar with the timestamp revoked portions of the keys as I don't believe this ticket/work has touched that, so I've left gpiiKey as is.

sgithens commented:
@cindyli I believe I've addressed all your comments, but am still working now on merging in the big master changes with the testing infrastructure couchdb changes. Hopefully should finish that here shortly.

sgithens commented:
@cindyli I think this should be ready for another round of review now. Thanks!

cindyli commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -73,12 +125,231 @@ var fluid = fluid || require("infusion");
+                    promiseTogo.reject({
+                        isError: true,
+                        message: "Cannot lookup prefsSafe",
+                        statusCode: 404
```
The return of "statusCode" at this level of code might be too early. Shall we leave it to request handlers?

This comment applies to other places.

sgithens replied:
The `statusCode` entries should all be in handlers now.

cindyli commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -73,12 +125,231 @@ var fluid = fluid || require("infusion");
+                if ($.isEmptyObject(data) || data.type !== "prefsSafe") {
+                    promiseTogo.reject({
+                        isError: true,
+                        message: "Cannot lookup prefsSafe",
```
With recent changes to the preferences server, a collection of error messages with "errorCode" is defined at [gpii.preferencesServer.errors[]](https://github.com/GPII/universal/blob/master/gpii/node_modules/preferencesServer/src/preferencesServerConst.js#L26) and shared across preferencesServer module. Could you follow the same pattern for error handling?

BTW, it would be helpful to think about if some similar messages can be replaced by one common message. For example, "Cannot lookup prefsSafe" here and "Error fetching prefsSafe" at line 207 probably can share one of these?

sgithens replied:
I've moved all of these into the errors, and am trying to reuse a few of them with when possible, although many require a specific check or comment that is unique.

cindyli commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -219,6 +486,59 @@ var fluid = fluid || require("infusion");
+        }
+        else {
+            // Preferences Safe must exist to add this key
+            that.getPrefsSafe(keyData.prefsSafeId).then(function () {
```
This else block seems a good use case of `fluid.promise.fireTransformEvent()`?

@amb26, I'm also confused about when `fluid.promise.fireTransformEvent()` should be used and when a chain of then() should be used. Your advice is appreciated.

sgithens replied:
It's interesting.  I think it makes sense for grouping together distinct operations... but it can be a little cumbersome when you need to add a precondition such as just checking for the existence of a prefssafe before continuing forward.

Maybe improving the tooling and streamlining the syntax for these chained operations and preconditions could be a BOF in Toronto next month.

amb26 replied:
I agree that this in-function promise chain is too long and should be broken up

cindyli commented:
Please update [.nycrc](https://github.com/GPII/universal/blob/master/.nycrc) to include new files you added into gpii/node_modules directory. Thanks.

cindyli commented on gpii/node_modules/preferencesServer/test/prefsSafesTests.js:
```diff
@@ -0,0 +1,715 @@
+            getRequest: {
+                type: "kettle.test.request.http",
+                options: {
+                    path: "/prefssafe-with-keys/" + fixture.url,
```
Can you write a test for this endpoint that a preferences safe only associates with one gpii key rather than more than one keys. I'm getting this error in this case:
```
FATAL ERROR: Uncaught exception: Cannot read property 'rows' of undefined
```

cindyli replied:
Actually, even if a preferences safe associates with more than one "gpiiKey" documents (without having any "gpiiCloudSafeCredential" documents), the same error still occurs.

sgithens replied:
Ok, the `FATAL ERROR` is definately something to be fixed, and you noted it below as well. However, I do believe that with the internal API it should be possible to associate more than one gpiiKey with a preferences set, since we have the ability to revoke and create new ones, and theoretically support having more than one key to a prefssafe.

cindyli replied:
Sorry that my original point was not clear enough.

I totally agree that one prefs safe can associate with multiple GPII keys. What I meant was, when I tested this endpoint, when a prefs safe associated with one or more GPII keys (in the database) without associating with any login credentials, the error mentioned above was hit.

cindyli commented on gpii/node_modules/preferencesServer/test/prefsSafesTests.js:
```diff
@@ -0,0 +1,715 @@
+/////////////////////// TESTING prefsSafes PUT (Save) ////////////////////////////
+fluid.registerNamespace("gpii.tests.preferencesServer.prefsSafes.putSafe");
+
+gpii.tests.preferencesServer.prefsSafes.putSafe.buildTestDef = function (fixture) {
```
A couple of issues with this endpoint. Please also write tests to cover these cases:

1. Sending this request with an non-existent prefs safe id will have the system auto create a new prefs safe with the given id. Is this what you want? The documentation says this endpoint is for updating existing safes. Please fix either the code or the documentation.
2. The code doesn't check if the prefsSafeId sent along as a request parameter matches the prefsSafeId in the PUT body. If there's a mismatch, the system still goes ahead to update the database with the PUT body.

sgithens replied:
I've updated the docs that this endpoint is _only_ for updating, and have added checks and unit tests for both of the above scenerios.

cindyli commented on gpii/node_modules/preferencesServer/test/prefsSafesTests.js:
```diff
@@ -0,0 +1,715 @@
+            getRequest: {
+                type: "kettle.test.request.http",
+                options: {
+                    path: "/prefssafes",
```
To re-produce the error I ran into:

1. Open a terminal, in the /universal root directory, run `NODE_ENV=gpii.config.preferencesServer.standalone.development.manualTesting.json5 node gpii.js` to start the prefs server;
2. Use curl command or other tools to send a GET request to http://localhost:8081/prefssafes;
3. The response received is:
```
{
    "isError": true,
    "message": "Cannot read property 'rows' of undefined"
}
```
Meanwhile, GPII started in the terminal terminates with this error:
```
FATAL ERROR: Uncaught exception: Cannot read property 'rows' of undefined
TypeError: Cannot read property 'rows' of undefined
    at gpii.dbOperation.dbDataStore.cleanUpDocList (/Users/gpii/node_modules/universal/gpii/node_modules/gpii-db-operation/src/DbDataStoreUtils.js:159:14)
```

sgithens replied:
This seems to be working for me now. I also tested going in to fauxton and deleting all the prefssafes to test with zero rows.  Your couch views might need to get to recreated?

cindyli commented on gpii/node_modules/preferencesServer/test/prefsSafesTests.js:
```diff
@@ -0,0 +1,715 @@
+        name: "POST: Add a new GPII Key to a preferences set",
+        postBody: {
+            "prefsSafeId": "prefsSafe-1",
+            "prefsSetId": "gpii-lowlight"
```
If the prefs set id "gpii-lowlight" doesn't exist in the "prefsSafe-1", this endpoint still goes ahead to create a new key that points to this non-existent prefs set. In this case, shall we either reject the request or create this prefs set with empty preferences in the prefs safe?

sgithens replied:
I'm a little hesitant to add this requirement at the moment... just because there are still a lot of changes going on with requirements, pilots, etc.  And because this API is internal I'd like to maybe leave this flexibility for now, in case we wanted to use the PPT or a script to batch create keys for some future preference safe / contextual situation ( I used the word context!!!)

Since we also don't really even support using prefsafe outside of default for now, I could imagine we might have use cases where you fall back to the default if you try to key in with a non-existent prefsset. I'm not sure yet, but it seems that things change enough, that we might want to wait on requiring this?

cindyli replied:
Well, sorry I disagree. I don't see a point in opening holes letting unsupported or mismatched data go into the system in order to give ways to some unknown flexibilities. I'd lean towards rejecting this kind of requests.

cindyli replied:
bump. I'm still not sure if we should create a GPII key if the "prefsSetId" it points to doesn't exist.

@amb26, what do you think?

the-t-in-rtf replied:
Based on my own work and previous conversations with @amb26, I wanted to chime in.  Since CouchDB cannot reasonably enforce this kind of record-to-record "linkage", we are left with a few strategies:

1. Screening for invalid external references when creating linkages.
2. Screening for invalid references when retrieving data (following linkages).
3. Checking for and cleaning up broken linkages with scripts.

 In my own work I have used all three approaches.  Although we should discuss, my gut feel is that if we don't at least check references when creating linkages (point 1) then we'll end up adding checks as odd, hard-to-troubleshoot bugs caused by broken linkages show up (point 2).  Anyway, will think on it further, it is indeed a good topic to discuss next Tuesday.

amb26 replied:
While I think a "soft consistency" approach is compatible with our general direction in a few areas (e.g. we now support requests referring to prefs safes that don't exist at all), I don't think it's reasonable to allow soft consistency in areas where we don't also supply adequate means for resolving it (e.g. creating the prefs safe that doesn't exist). So I agree we should throw an error in this case unless we are prepared to do something right now to resolve the inconsistency (e.g. create the nonexistent prefs set)

cindyli commented:
Please add documentation for `PUT /add-cloud-credentials/:prefsSafeId` endpoint.

cindyli commented on gpii/node_modules/preferencesServer/test/cloudSafeCredTests.js:
```diff
@@ -0,0 +1,375 @@
+            putCloudCred: {
+                type: "kettle.test.request.http",
+                options: {
+                    path: "/add-cloud-credentials/" + fixture.url,
```
Using the same steps described [here](https://github.com/GPII/universal/pull/634/#discussion_r266045546) to send PUT request to `http://localhost:8081/add-cloud-credentials/prefsSafe-carla` to reproduce these issues. Please write a test to cover these cases too:

1. When not providing PUT body or providing PUT body that does not have "password" field in the object, GPII terminates with this error:
```
FATAL ERROR: Uncaught exception: Pass phrase must be a buffer
TypeError: Pass phrase must be a buffer
    at pbkdf2 (crypto.js:702:20)
    at Object.exports.pbkdf2Sync (crypto.js:687:10)
...
```

2. Add credential for a prefs safe that never has a credential record by providing PUT body with "password" without "username" such as:
```
{
	"password": "bbb"
}
```
Receives this reponse:
```
{
    "isError": true,
    "message": "Unable to create credentials with that username and password combination"
}
```
Meanwhile, GPII console log shows:
```
Error adding cloud safe credential{
    "isError": true,
    "message": {
        "error": "conflict",
        "reason": "Document update conflict."
    }
}
```
`Document update conflict` error is odd because there has never been a credential record created for this prefs safe.

cindyli replied:
If the given username is an empty string or a couple of spaces, `PUT /add-cloud-credentials` endpoint will still go ahead to create these usernames. There should be stronger sanitization.

sgithens replied:
I've added unit tests and validation to make sure the payload is basically sane, has both fields and they aren't empty.  There is somewhat extensive rules being developed by UX and requirements for what should be considered allowable for deployments, and would like to leave the work for those more complex rules to another jira once the requirements are locked down.

sgithens commented:
@cindyli I believe this should be ready for another review now.  All the comments should be addressed except for the few above I left notes on.

cindyli commented on gpii/node_modules/gpii-db-operation/src/DbUtils.js:
```diff
@@ -36,3 +36,40 @@ gpii.dbOperation.composeError = function (error, termMap) {
+ * The current criteria for validity is that the item is not undefined and has
+ * the `type` that is supplied in the arguments.
+ *
+ * @param  {Object} item - Item, mostly a return from another API or CouchDB call
```
Please remove the extra space in between "@param" and "{". Can you do a search with "@param  " and fix them all in the repo. Thanks.

sgithens replied:
updated

cindyli commented on gpii/node_modules/gpii-db-operation/src/DbUtils.js:
```diff
@@ -36,3 +36,40 @@ gpii.dbOperation.composeError = function (error, termMap) {
+ * included in the promise rejection payload.
+ * @param  {String} errorCode - If this isn't the appropriate type, this error code will be
+ * included in the promise rejection payload.
+ * @return {fluid.promise} Promise, when the item is valid is resolves with the
```
is resolved -> , it resolves

sgithens replied:
updated

cindyli commented on gpii/node_modules/gpii-db-operation/test/DbDataStoreTests.js:
```diff
@@ -510,6 +510,54 @@ fluid.defaults("gpii.tests.dbDataStore.addAuthorization", {
+                task: "{dbDataStore}.findRelatedDocsForPrefsSafe",
+                args: ["prefsSafe-1"],
+                resolve: "jqUnit.assertLeftHand",
+                resolveArgs: ["Expecing a list of keys for prefsSafe-1.", gpii.tests.dbDataStore.testData.findRelatedDocsForPrefsSafe_prefsSafe1, "{arguments}.0"]
```
Expecing -> Expecting, missing "t".

sgithens replied:
updated

cindyli replied:
Same typo at line 523.

cindyli commented on gpii/node_modules/gpii-db-operation/test/DbDataStoreTests.js:
```diff
@@ -510,6 +510,54 @@ fluid.defaults("gpii.tests.dbDataStore.addAuthorization", {
+                task: "{dbDataStore}.findSafeByExpressUserLookup",
+                args: ["org.couch.db.user:prefs1user"],
+                resolve: "jqUnit.assertDeepEq",
+                resolveArgs: ["Expecing a safe from the express lookup.", gpii.tests.dbDataStore.testData.findSafeByExpressUserLookup_credential, "{arguments}.0"]
```
same as above. typo "Expecing".

sgithens replied:
updated

cindyli commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -73,12 +124,270 @@ var fluid = fluid || require("infusion");
+     * is returned.
+     *
+     * @param  {Object} dataStore - Instance of `gpii.dbOperation.dataStore`
+     * @param  {String} prefsSafeId Prefs Safe ID
```
Fix the jsdoc format at line 184, 185. also remove the extra space in btw "@param" and "{".

sgithens replied:
updated

cindyli commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -73,12 +124,270 @@ var fluid = fluid || require("infusion");
+
+    /**
+     * Adds a new preferences safe to CouchDB using the preferences safe format
+     * minus the `id`, `timestampCreated`, and `timestampUpdated` properties.
```
This statement of "minus the `id`, `timestampCreated`, and `timestampUpdated` properties" is misleading. It was the input `prefsSafeData` value does not require these fields, but this function will end up with creating all these fields in the database. Probably use this statement for `@param {Object} prefsSafeData` and re-describe the function goal here.

sgithens replied:
updated

cindyli commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -73,12 +124,270 @@ var fluid = fluid || require("infusion");
+            });
+        }
+        else {
+            dataStore.addPrefsSafe(prefsSafeData).then(
```
Add a todo here that "prefsSafeData" should be validated when schema validation is implemented. The same comment applies to other functions that create or update documents.

sgithens replied:
updated

cindyli commented on documentation/PreferencesServer.md:
````diff
@@ -350,6 +350,434 @@ There are two important things to note here:
+```
+
+A successful return payload will consist of the same payload with the addition of timestamps and a
+freshly generated `id` field.  In the event of an error a payload similar to the following will be returned
````
Missing comma after "In the event of an error".

It would be nice to add a small section to describe what error cases are covered. This comment applies to other places where this generic system error description is used. ;)

sgithens replied:
updated

cindyli commented on documentation/PreferencesServer.md:
```diff
@@ -350,6 +350,434 @@ There are two important things to note here:
+### GET /prefssafe-keys/:prefsSafeId
+
+This will return the associated keys and credentials documents for a given preferences safe, in a `rows` field.
+Also included is a `total_rows` and `offset` field, but do note that the `total_rows` field is not accurate as
```
What's `offset` field for? Is there a reason that `total_rows` value is left inaccurate?

sgithens replied:
`offset` is a standard couch field that indicated how far from the beginning of the dataset your first item is.  This is allow for paging if you are only getting a subset of the return data.  

I can't remember why `offset` doens't actually work the way you would expect in couch. I think @the-t-in-rtf explained it to me once.  It might be a good thing for us to address as a whole project wise as we'll have more and more endpionts returning arrays.  For now, I've just documented it as such.

cindyli replied:
Just out of curiosity, I did some search and experiment to see what's going on with `total_rows`. It turns out this field seems returning the correct number it's supposed to return.

[The CouchDB doc](http://docs.couchdb.org/en/stable/api/ddoc/views.html#db-design-design-doc-view-view-name) says `totoal_rows` holds "Number of documents in the database/view". This endpoint uses the view "findRelatedDocsForPrefsSafe" so `total_rows` should have the value of total number of documents in this view.

If you do a CouchDB query on this view without providing a specific `key` value: `/gpii/_design/views/_view/findRelatedDocsForPrefsSafe?include_docs=true`, the query returns all documents in this view, whose total matches the value of `total_rows` when this view is queried with a specific key.

If you could double confirm on this, the statement of "the `total_rows` field is not accurate" can be removed from here.

cindyli commented on documentation/PreferencesServer.md:
```diff
@@ -350,6 +350,434 @@ There are two important things to note here:
+### POST /prefssafe-key-create
+
+This endpoint will add a new `gpiiKey` document to the system. There are no URL parameters, but the JSON body takes
+3 fields, one of them optional. Required are the `prefsSafeId` and `prefsSetId`. These indicate the preferences safe,
```
"one of them is optional." Missing "is".

sgithens replied:
updated

cindyli commented on documentation/PreferencesServer.md:
```diff
@@ -350,6 +350,434 @@ There are two important things to note here:
+
+This endpoint will add a new `gpiiKey` document to the system. There are no URL parameters, but the JSON body takes
+3 fields, one of them optional. Required are the `prefsSafeId` and `prefsSetId`. These indicate the preferences safe,
+and the respective preferences set that this key and token will operate upon. Optionally you can pass in a unique
```
"this key and token" -> "this GPII key". All appearances of "user token", "token" that reference a GPII key should be replaced with "GPII key". There had been a round of terminology sync in this repo a few months ago. This comment applies to the entire pull request.

sgithens replied:
updated

cindyli commented:
I ran into a problem with testing POST /unlock-cloud-safe endpoint:

1. Start the standalone preferences server by running this command in a terminal at /universal root directory: `NODE_ENV=gpii.config.preferencesServer.standalone.development.manualTesting.json5 node gpii.js`

2. add a cloud credential by sending PUT request to `http://localhost:8081/add-cloud-credentials/prefsSafe-carla with the PUT body:
```
{
    "username": "NewLoginUsername",
    "password": "v3ry$ecu4e"
}
```
A successful response is received.

3. unlock the cloud credential by sending a POST request to `http://localhost:8081/unlock-cloud-safe` with the POST body same as the PUT body at step 2. Received an error response:
```
{
    "isError": true,
    "message": "Unable to unlock Preferences Safe with the supplied credentials.",
    "errorCode": "GPII_ERR_UNLOCKING_PREFSSAFE_CREDENTIALS"
}
```
I expect to receive a successful response at step 3.

Let me know if I did something wrong in these steps.

sgithens commented:
@cindyli I just tested this manually using a RESTful API tool and was able to add and unlock a credential that I added to `prefsSafe-1`. I also tested it with a bad password to make sure it would fail.

In my test data there isn't a prefsSafe-carla, but I also tried adding credentials to an non existing safe, and got the correct error that I couldn't add it to a safe that doesn't exist.  I'm not sure what you might be running in to without more details.

> I ran into a problem with testing POST /unlock-cloud-safe endpoint:
> 
> 1. Start the standalone preferences server by running this command in a terminal at /universal root directory: `NODE_ENV=gpii.config.preferencesServer.standalone.development.manualTesting.json5 node gpii.js`
> 2. add a cloud credential by sending PUT request to `http://localhost:8081/add-cloud-credentials/prefsSafe-carla with the PUT body:
> 
> ```
> {
>     "username": "NewLoginUsername",
>     "password": "v3ry$ecu4e"
> }
> ```
> A successful response is received.
> 
> 1. unlock the cloud credential by sending a POST request to `http://localhost:8081/unlock-cloud-safe` with the POST body same as the PUT body at step 2. Received an error response:
> 
> ```
> {
>     "isError": true,
>     "message": "Unable to unlock Preferences Safe with the supplied credentials.",
>     "errorCode": "GPII_ERR_UNLOCKING_PREFSSAFE_CREDENTIALS"
> }
> ```
> I expect to receive a successful response at step 3.
> 
> Let me know if I did something wrong in these steps.

sgithens commented:
ok to test

sgithens commented:
This should be ready for another round of review and hopefully approval for merge. (even if it needs to wait till after the pilot is deployed)

cindyli commented on documentation/PreferencesServer.md:
````diff
@@ -350,6 +350,434 @@ There are two important things to note here:
+```
+
+A successful return payload will consist of the same payload with the addition of timestamps and a
+freshly generated `id` field.  In the event of an error, a payload similar to the following will be returned
````
"In the event of an error"... what is this error? Could you describe actual errors like what you did above at line 461 rather than using an fabricated error code ""GPII_ERR_APPLICABLE_MESSAGE"? Thanks.

This comment applies to other places in this doc.

sgithens replied:
I went back through each `isError` message doc in the markdown. In situations where I'm actually doing specific checks I've put in the potential items to look for.  The remainders of them are actual general error catches for anything potentially unexpected, which is why they are general error messages, but do include the message.

cindyli commented on gpii/node_modules/gpii-db-operation/src/DbUtils.js:
```diff
@@ -36,3 +36,40 @@ gpii.dbOperation.composeError = function (error, termMap) {
+ * included in the promise rejection payload.
+ * @param {String} errorCode - If this isn't the appropriate type, this error code will be
+ * included in the promise rejection payload.
+ * @return {fluid.promise} Promise, when the item is valid it resolves with the
```
Missing coma after "when the item is valid".

sgithens replied:
updated

cindyli commented:
Regarding the problem I had with not being able to unlock a credential, the debug shows it's because [the views json file from gpii-express-user](https://github.com/GPII/gpii-express-user/blob/master/src/views/lookup.json) is not loaded into the database so the query on the view "byUsernameOrEmail" returns missing error.

I noticed [a duplicate of this view json file](https://github.com/sgithens/universal/blob/GPII-2966/gpii/node_modules/preferencesServer/test/data/userLookup.json) is used for preference server tests. Do you think it makes sense to replace this strategy with using the real json from gpii-express-user module for both tests and manual testing configs. Thanks.

cindyli commented:
Just a side note as a reminder: Let's talk about [this](https://github.com/GPII/universal/pull/634#pullrequestreview-215132109) at the meeting next Tuesday.

sgithens commented:
ok to test

amb26 commented on gpii/node_modules/preferencesServer/src/preferencesServer.js:
```diff
@@ -52,6 +52,52 @@ fluid.defaults("gpii.preferencesServer", {
+        },
+        // GPII-2966
+        prefsSafeGet: {
+            route: "/prefssafe/:prefsSafeId",
```
"prefsSafe" for all these endpoints would be consistent with our usage everywhere else

amb26 replied:
Bump comment

sgithens replied:
updated

sgithens commented:
@cindyli @amb26 @klown Hi everyone. This should be ready for another review now.  To recap, this branch contains the initial work necessary for username/password account credentials. Previously Cindy had reviewed the entire branch and ok'd it, but the remaining item was not having a migration to remove the `password` field from prefsSafes. (Which isn't actually necessary for it to operate but it's nice to clean up.)

For the migration, I've largely cloned the previous GPII-4011 migration Joseph did. Since we're going to be migrating from couchdb shortly I didn't spend a lot of time trying to refactor the common code from the 2 migrations since it seems like it won't be necessary in the future.

cindyli commented:
> Regarding the problem I had with not being able to unlock a credential, the debug shows it's because [the views json file from gpii-express-user](https://github.com/GPII/gpii-express-user/blob/master/src/views/lookup.json) is not loaded into the database so the query on the view "byUsernameOrEmail" returns missing error.
> 
> I noticed [a duplicate of this view json file](https://github.com/sgithens/universal/blob/GPII-2966/gpii/node_modules/preferencesServer/test/data/userLookup.json) is used for preference server tests. Do you think it makes sense to replace this strategy with using the real json from gpii-express-user module for both tests and manual testing configs. Thanks.

This comment doesn't seem to be addressed.

sgithens commented:
@cindyli Ok, here is what I've done, with a little explanation. 

I removed the custom userLookup.json views from the branch.  I had on previous occasions thought also that it would be better to use the actual file from gpii-express-user, except that gpii-express-user uses a slightly different method of loading it's data.  It's json views are in a single top level JSON object in the file, where-as ours are contained within a top level JSON array.  So our dataloader won't load it. I had looked at possibly just loading and transforming the data, expect that the dataloader many layers down expects a file name rather than the JSON, and it would be quite a bit of refactoring to change this I think.

Given that we are planning on shortly moving off of couchdb, it didn't seem as if this was worth my time, so I have just included the necessary views along ours in the views.json. I also think this makes the provisioning code simpler, as it works as it, and I don't need to fix the dataloader to load more view files from other node_modules.   If we weren't in our last days with couch I would say maybe we should work more on loading views from 3rd party modules but it doesn't seem worth it right now.

Also... based on our in flux requirements for user accounts and authentication, I can see us actually wanting to customize these couchdb views slightly for universal. For instance we might want to change the byUsernameOrEmail view a bit to tighten up how the credentialLogin documents are found. I'm not sure if we'll do that, but I wouldn't be surprised if we did.

Anyways, this is not perfect, but it's very simple, doesn't change any of our deployment scripts, and I think would work best for our current database trajectory.

cindyli commented:
Thanks for the explanation, @sgithens. I'm fine with this approach. As you said, we're moving off of CouchDB soon.

amb26 commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -73,12 +124,273 @@ var fluid = fluid || require("infusion");
+                // TODO This will have json schema validation when the scaffolding is all in.
+                dataStore.updatePrefsSafe(prefsSafeData.id, prefsSafeData).then(
+                    function (/* updatedReturn */) {
+                        that.getPrefsSafe(prefsSafeData.id).then(
```
This is a too deeply nested promise chain and needs to be broken up

sgithens replied:
This has been converted to a transforming promise chain.

amb26 commented on gpii/node_modules/preferencesServer/src/preferencesService.js:
```diff
@@ -16,22 +16,58 @@ var fluid = fluid || require("infusion");
                 type: "gpii.dbOperation.dataStore"
+            },
+            expressUserUtils: {
+                type: "gpii.express.user.utils",
```
I am very suspicious of a component named "utils" ...... what does it do?

sgithens replied:
This is a component in the `gpii-express-user` module that contains the routines for creating and unlocking users.

https://github.com/GPII/gpii-express-user/blob/010bbaf83ae2e43558878f77b43e13c60223a0e3/src/js/server/utils.js#L29

amb26 commented:
Please could you try out merging this against https://github.com/GPII/universal/pull/833 to flush out whether it happens to perform any modification of component options which will be prohibited under the new framework

sgithens commented:
@amb26 I've address the couple comments above.  I'm investigating a test error when merging in your FLUID-6145 branch, which may or may not be related to gpii-express-user.  But in general I believe this is ready for another review.
